kaggle_environments/envs/lux_ai_2022/luxai2022/env.py

```diff
@@ -0,0 +1,618 @@
+            for agent in self.agents:
+                if failed_agents[agent]:
+                    continue
+                for unit in self.state.units[agent].values():
+                    unit_a: Action = unit.next_action()
+                    if unit_a is None:
+                        continue
+                    actions_by_type[unit_a.act_type].append((unit, unit_a))
+                for factory in self.state.factories[agent].values():
+                    if len(factory.action_queue) > 0:
+                        unit_a: Action = factory.action_queue.pop(0)
+                        actions_by_type[unit_a.act_type].append((factory, unit_a))
+
+            for agent in self.agents:
+                for factory in self.state.factories[agent].values():
+                    # update information for lichen growing and cache it
+                    factory.cache_water_info(self.state.board, self.env_cfg)
+
+            # 3. validate all actions against current state, throw away impossible actions TODO
+            actions_by_type = validate_actions(self.env_cfg, self.state, actions_by_type, verbose=self.env_cfg.verbose, weather_cfg=weather_cfg)
+            # TODO test Transfer resources/power
+
+            for unit, transfer_action in actions_by_type["transfer"]:
+                transfer_action: TransferAction
+                transfer_amount = unit.sub_resource(transfer_action.resource, transfer_action.transfer_amount)
+                transfer_pos: Position = unit.pos + move_deltas[transfer_action.transfer_dir]
+                units_there = self.state.board.get_units_at(transfer_pos)
+
+                # if there is a factory, we prefer transferring to that entity
+                factory_id = f"factory_{self.state.board.factory_occupancy_map[transfer_pos.y, transfer_pos.x]}"
+                if factory_id in self.state.factories[unit.team.agent]:
+                    factory = self.state.factories[unit.team.agent][factory_id]
+                    # add resources to target. This will waste (transfer_amount - actually_transferred) resources
+                    actually_transferred = factory.add_resource(transfer_action.resource, transfer_action.transfer_amount)
+                elif units_there is not None:
+                    assert len(units_there) == 1
+                    target_unit = units_there[0]
+                    # add resources to target. This will waste (transfer_amount - actually_transferred) resources
+                    actually_transferred = target_unit.add_resource(transfer_action.resource, transfer_amount)
+
+            for unit, pickup_action in actions_by_type["pickup"]:
+                pickup_action: PickupAction
+                factory = self.state.board.get_factory_at(unit.pos)
+                pickup_amount = factory.sub_resource(pickup_action.resource, pickup_action.pickup_amount)
+                # may waste resources if tried to pickup more than one can hold.
```

Comment:
do we call this out in the specs? if you try to pickup more than you can take the excess is wasted rather than just staying in the factory? Seems like it would be reasonable/beginner friendly to only limit pickup to how much they can take?

Reply:
I tried to stay consistent for the unit-to-unit transfers where unmanaged transferring can lead to resource waste (e.g. 4 units transferring a total amount of water greater than `cargo_space` to the same unit)